spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/TestFunctionCatalog.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.iceberg.AssertHelpers;
+import org.apache.iceberg.IcebergBuild;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.spark.functions.SparkFunctions;
+import org.apache.spark.sql.AnalysisException;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.analysis.NoSuchFunctionException;
+import org.apache.spark.sql.catalyst.analysis.NoSuchNamespaceException;
+import org.apache.spark.sql.connector.catalog.FunctionCatalog;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.catalog.functions.ScalarFunction;
+import org.apache.spark.sql.connector.catalog.functions.UnboundFunction;
+import org.apache.spark.sql.types.StructType;
+import org.assertj.core.api.Assertions;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestFunctionCatalog extends SparkCatalogTestBase {
```

Comment:
Can you run this with just one catalog? 3 seems unnecessary.

Reply:
Sure.

I wanted to test session catalog vs non session catalog as non-session catalog is somewhat different, mostly because calling functions in SQL on the session catalog uses v1 behavior and ensures the namespace exists.

But I’ve added a check if in session catalog and create `session` namespace block and will leave that in as documentation & run then run this code as session catalog.

3 is too many for sure.

Reply:
@kbendick, can you update this test?

Reply:
Yeah. I reverted it from 1 back to 3 momentarily to test refactoring this class. I'll revert it back once I'm done and ping you again. Sorry for the confusion.

Reply:
Ok it's back at just one.

A lot of the `FunctionCatalog` / v2 functions cannot use many DDL statements like `SHOW FUNCTIONS` and `DESCRIBE FUNCTION`.

This is in-line with what's in the V2CatalogSuite in Spark as well, so I've added the tests via just calling the code directly.

Given that's generally how Spark uses the function catalog functions, in addition to resolving them for usage in SQL, that is the best we can do.

However, I scrapped the old large test and broke it up so it should be much more readable now.